src/IssueViz/Editor/FilteringTaggerBase.cs

```diff
@@ -0,0 +1,175 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Tagging;
+
+namespace SonarLint.VisualStudio.IssueVisualization.Editor
+{
+    /// <summary>
+    /// Base class for a tagger that produces tags that are based on another tag type
+    /// e.g. a tagger that produces IErrorTags based on our IIssueLocationTags
+    /// </summary>
+    /// <typeparam name="TTrackedTagType">The type of tag being tracked</typeparam>
+    /// <typeparam name="TTagType">The type of tag being produced</typeparam>
+    internal abstract class FilteringTaggerBase<TTrackedTagType, TTagType> : ITagger<TTagType>, IDisposable
```

Comment:
We will have multiple taggers inheriting from this base class:
* a buffer tagger that produces `IErrorTags` by filtering `IIssueLocationTags` (i.e. the `ErrorTagger` in this PR)
* a buffer tagger that tracks the active selected issue and produces `ISecondaryLocationTags`
* a view tagger that produces highlights for the `ISecondaryLocationTags` and
* a view tagger that produces adornments for the `ISecondaryLocationTags`

Reply:
I don't think it has to be inheritance. If you take out `GetTags` method, there isn't a lot left to to inherit. 
Why not extract `GetTags` to a service and use it in each tagger as needed? It can accept the relevant methods as a callback (CreateTagSpan, Filter).

Reply:
It's a tricky one. I try to avoid inheritance, but it's useful here since there is a lot of common behaviour that would be more painful to due using composition i.e. event registration, handling, propagation and cleanup.
For me, the problem is how to isolate the variable bits of the sub-classes to make them easier to test, which means the new tag creation and the filtering, rather than `GetTags` itself. I considered a couple of alternatives, including separating out `GetTags` from the base class, but I didn't think that really helped as it didn't make the tag creation or filtering methods easier to test. I also considered a separate interface with `Filter` and `CreateTagSpan` methods, similar to your suggestion about using separate callbacks.
I'll look at it when adding the tagger classes to see if it makes things easier.